testflinger/api/v1.py

```diff
@@ -345,3 +376,37 @@ def job_position_get(job_id):
         ):
             return str(position)
     return "Job not found or already started\n", 410
+
+
+def cancel_job(job_id):
+    """Cancellation for a specified job ID
+
+    :param job_id:
+        UUID as a string for the job
+    """
+    result_file = os.path.join(current_app.config.get("DATA_PATH"), job_id)
+    if not os.path.exists(result_file):
+        return "Job is not found and cannot be cancelled\n", 400
+    try:
+        with open(
+            result_file, "r", encoding="utf-8", errors="ignore"
+        ) as results:
+            data = json.load(results)
+    except json.decoder.JSONDecodeError:
+        # If for any reason it's empty or has bad data
+        data = {"job_state": "bad_data"}
+    if data["job_state"] in ["complete", "cancelled"]:
+        return "The job is already completed or cancelled", 400
+    # Remove job from the queue if it hasn't been started
+    if data["job_state"] == "waiting":
```

Comment:
Ah, there's actually another state in which it would be valid to remove it - resubmitted. It's a little less common but can definitely happen.  Also, even if a job is running, we can cancel it! We won't be able to remove it from the redis list, but it's safe to call LREM on it even if it won't exist there (if that's easier).

I think it might be better to simply fail if the state is cancelled or complete. Any other state should be ok to try and cancel the job.